[MRG +1] adds DeprecationWarning to DPGMM and VBGMM

juliathebrave opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/master/CONTRIBUTING.md#Contributing-Pull-Requests
-->
#### Reference Issue

<!-- Example: Fixes #1234 -->

Fixes #6998
#### What does this implement/fix? Explain your changes.

I added a DeprecationWarning to the class DPGMM and the class VBGMM, as well as public functions in dpgmm.py

I added tests for these warnings in test_dpgmm.py
#### Any other comments?

@tguillemot 

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

juliathebrave committed:
added Deprecation Warnings to DPGMM and VBGMM classes and other functions

juliathebrave committed:
modified/added functions test_digamma, test_gammaln, test_log_normalize to check for DeprecationWarning

juliathebrave committed:
added functions to test for Deprecated Warnings for wishart_log_det, wishart_logz

juliathebrave committed:
updated test_log_normalize function

juliathebrave committed:
fixed spelling in DeprecationWarning for class DPGMM and
added DeprecationWarning tests test_DPGMM_deprecation, test_VBGMM_deprecation

juliathebrave committed:
added blank lines near inserted DeprecatedWarnings to conform to pep8

juliathebrave committed:
added blank lines and deleted trailing whitespace for pep8

juliathebrave committed:
removed 'import warnings' since not used

tguillemot commented on sklearn/mixture/dpgmm.py:
```diff
@@ -16,21 +16,27 @@
 
 
+@deprecated("The function digamma is deprecated in 0.18 and "
+            "will be removed in 0.20.")
```
Can you add "Use `scipy.digamma` instead."

tguillemot commented on sklearn/mixture/dpgmm.py:
```diff
@@ -16,21 +16,27 @@
 
 
+@deprecated("The function gammaln is deprecated in 0.18 and "
+            "will be removed in 0.20.")
```
Can you add "Use `scipy.special.gmmaln instead.`"

tguillemot commented on sklearn/mixture/dpgmm.py:
```diff
@@ -106,6 +116,8 @@ def _bound_state_log_lik(X, initial_bound, precs, means, covariance_type):
 
 
+@deprecated("The class DPGMM is deprecated in 0.18 and "
+            "will be removed in 0.20. This class no longer works.")
```
Can you change the message in that way "The DPGMM class is not working correctly and it's better to not use it. DPGMM is deprecated in..."

amueller replied:
I'm not sure about "this class no longer works". I think we should say explicitly that it will provide bad results. Maybe ``This class is broken and may yield bad results"?

tguillemot commented on sklearn/mixture/dpgmm.py:
```diff
@@ -606,6 +618,8 @@ def _fit(self, X, y=None):
 
 
+@deprecated("The class VBGMM is deprecated in 0.18 and "
+            "will be removed in 0.20. This class no longer works.")
```
Can you change the message in that way "The VBGMM class is not working correctly and it's better to not use it. DPGMM is deprecated in..."

dhanus commented:
look good to me. +1

tguillemot commented on sklearn/mixture/tests/test_dpgmm.py:
```diff
@@ -89,10 +93,48 @@ def test_verbose_second_level():
 
+def test_digamma():
+    assert_warns_message(DeprecationWarning, "The function digamma is"
+                         " deprecated in 0.18 and will be removed in 0.20.",
```
Change the message accordingly

tguillemot commented on sklearn/mixture/tests/test_dpgmm.py:
```diff
@@ -89,10 +93,48 @@ def test_verbose_second_level():
+
+def test_gammaln():
+    assert_warns_message(DeprecationWarning, "The function gammaln"
+                         " is deprecated in 0.18 and will be removed"
```
Same as before

juliathebrave replied:
If you look at the code by pressing 'View full changes', you'll see that the phrase 'Use scipy.special.digamma instead.' has been added.

tguillemot commented on sklearn/mixture/tests/test_dpgmm.py:
```diff
@@ -89,10 +93,48 @@ def test_verbose_second_level():
+def test_DPGMM_deprecation():
+    assert_warns_message(DeprecationWarning, "The class DPGMM is"
+                         " deprecated in 0.18 and will be removed in 0.20."
+                         " This class no longer works.", DPGMM)
```
Change the message accordingly

tguillemot commented on sklearn/mixture/tests/test_dpgmm.py:
```diff
@@ -128,6 +170,12 @@ class TestDPGMMWithFullCovars(unittest.TestCase, DPGMMTester):
+def test_VBGMM_deprecation():
+    assert_warns_message(DeprecationWarning, "The class VBGMM is"
+                         " deprecated in 0.18 and will be removed in "
+                         "0.20. This class no longer works.", VBGMM)
```
Same as before.

tguillemot commented:
@juliathebrave Thanks.
Have you removed the warning message when you do `nosetests -v` in the mixture repo.

amueller commented:
I'm not sure ``DeprecationWarning`` is the right warning for this ... hum.

amueller commented:
I guess deprecation warning is fine.

juliathebrave committed:
changed wording of DeprecationWarnings for style and/or recommend alternate functions

juliathebrave committed:
fixed spelling error in warning for gammaln

amueller commented:
LGTM.

juliathebrave commented:
@tguillemot No DeprecationWarnings from this branch show up in nosetests -v

juliathebrave committed:
suppress DecprecationWarnings for two functions

juliathebrave committed:
fixed ignore_warnings function usage

juliathebrave committed:
added more @ignore_warnings(DeprecationWarning)

tguillemot commented:
There are still some warnings which appears during the process.

juliathebrave committed:
temporary commit to check nosetests on travis

juliathebrave committed:
TST suppresses more DeprecatedWarnings

addresses GH6998 VBGMM and DPGMM deprecation

juliathebrave commented:
Most of the DeprecatedWarnings are gone. None show up on my computer. I counted 4 in Travis (2 GMM and 2 from DPGMM/VBGMM).

juliathebrave committed:
CLN fixes whitespace for pep8

in test_gmm.py and test_dpgmm.py
for GH6998

amueller commented on sklearn/mixture/tests/test_dpgmm.py:
```diff
@@ -128,6 +182,14 @@ class TestDPGMMWithFullCovars(unittest.TestCase, DPGMMTester):
     setUp = GMMTester._setUp
 
 
+@ignore_warnings(category=DeprecationWarning)
```
you don't actually need these because ``assert_warns_message`` already catches the deprecation, IIRC.

amueller commented on sklearn/mixture/tests/test_gmm.py:
```diff
@@ -315,16 +324,16 @@ class TestGMMWithDiagonalCovars(unittest.TestCase, GMMTester):
-    covariance_type = 'tied'
-    model = mixture.GMM
-    setUp = GMMTester._setUp
+# class TestGMMWithTiedCovars(unittest.TestCase, GMMTester):
```
Why did you comment these out?

amueller commented on sklearn/mixture/tests/test_gmm.py:
```diff
@@ -216,46 +217,51 @@ def test_sample(self, n=100):
         g.covars_ = np.maximum(self.covars[self.covariance_type], 0.1)
         g.weights_ = self.weights
 
-        samples = g.sample(n)
```
I guess this is fine, but it might have been easier / shorter to use a decorator on the whole function. But don't worry about it now.

amueller commented:
LGTM apart from the commented out stuff. I'd rather not remove the tests, but if they need to be removed, you need to delete them.

juliathebrave commented:
I'll check again soon regarding commented out sections -- if they were left commented out it was an oversight on my part. I was commenting to deduce what was throwing errors. 

Unfortunately, the decorator at the beginning of the function was not suppressing enough errors. That's the reason for the additional ignore_warnings code inside functions.

amueller commented:
That's surprising to me and probably a bug in the decorator. After this is finished, it might be interesting to investigate ;)

tguillemot commented:
@amueller I agree.

juliathebrave committed:
TST uncommented some code

I had commented some code for debugging purposes and had forgotten
to uncomment it. Now it is uncommented again.

amueller commented:
LGTM. @tguillemot ?

tguillemot commented:
LGTM
Thanks @juliathebrave

juliathebrave commented:
Thanks for the introduction to contributing to scikit-learn :) 
@tguillemot @amueller
